Improve Node Blocks

tn12787 opened the pull request:
address #586.

omichel commented:
I am not sure about the `[list of strings]`, `[any string]`, `[lookup table]` as it seems redundant with the `MFString`, `SFString` and `lookupTable` information present on the same line... I would simply make no comment in these cases.

stefaniapedrazzi commented:
Also note that a PROTO is a node and we would like that there won't be any difference between basic nodes and PROTO nodes for users, so I do not think it is good to specify it with `[node, PROTO]`.

tn12787 commented:
My reason for using the `[node, PROTO]` notation is primarily that other fields can admit either one specific node, or any PROTO instance based on that node - the `material` field of a `Shape` node can admit either a `Material` node or any `Material`-based PROTO.

I think it's important to put PROTOs at the forefront as they're powerful and actually very useful. I wonder whether more people would take the time to learn and use them if we highlight them more like this.

stefaniapedrazzi commented:
But they are all nodes.
A Material node is a node and a PROTO based on Material is a node.
It is good, to specific which specific kind of nodes can be inserted. But there to me it is useless and confusing to not treat PROTO as nodes but to specify them.

If we want to have a user-friendly and easy-to-use interface a users should not care if he is using a PROTO or a node. For him in the scene tree they should look exactly the same..
Moreover only advanced users will ever create PROTO nodes and need to really know the difference.

But we can better discuss this during the meeting.

omichel commented on reference/background.md:
```diff
@@ -2,13 +2,13 @@
+  MFString frontUrl  [ ]   # [list of strings]
+  MFString leftUrl   [ ]   # [list of strings]
+  MFString rightUrl  [ ]   # [list of strings]
+  MFString topUrl    [ ]   # [list of strings]
```
In that case, I believe we might be more precise, like: `# [list of image URLs]`

tn12787 replied:
This sounds good to me!

tn12787 commented:
I've fixed a couple of latent problems with the doc css here too.

tn12787 commented:
:+1:

omichel commented on reference/background.md:
```diff
@@ -2,13 +2,13 @@
+  MFString leftUrl   [ ]   # [list of Image URLs]
+  MFString rightUrl  [ ]   # [list of Image URLs]
+  MFString topUrl    [ ]   # [list of Image URLs]
+  MFColor  skyColor  [ 0 0 0 ]    # [0,1]
```
Missing space after coma.

omichel commented on reference/compass.md:
```diff
@@ -4,11 +4,11 @@ Derived from [Device](device.md).
+  SFBool  xAxis       TRUE   # [TRUE, FALSE]
+  SFBool  yAxis       TRUE   # [TRUE, FALSE]
+  SFBool  zAxis       TRUE   # [TRUE, FALSE]
+  SFFloat resolution  -1     # [-1, [0, inf]]
```
Woow, that's not obvious to understand... Maybe we should use the standard mathematical representation for sets and intervals, which would translate into: `{ -1, [0, inf) }`. If we use this scientific formalism, which I believe is good, we should also change all the `[TRUE, FALSE]` to `{TRUE, FALSE}` and `[0, inf]` to `[0, inf)`, `[ImageTexture, PROTO]` to `{ImageTexture, PROTO}`, `["symmetric", "active", "passive"]` to `{"symmetric", "active", "passive"}`, etc. We should also remove the square brackets around text describing strings to avoid confusion.

See explanations here: https://en.wikipedia.org/wiki/Set_notation

tn12787 commented:
how is this?

tn12787 commented:
👍

omichel commented on reference/recognition.md:
```diff
@@ -2,11 +2,11 @@
-  SFBool   occlusion      TRUE    # should occlusions between the camera and the objects be checked
-  SFColor  frameColor     1 0 0   # color of the object's frame in the overlay
-  SFInt32  frameThickness 1       # thickness of the object's frame in the overlay (pixels)
+  SFFloat  maxRange       100     # [0, inf]
```
wrong

omichel commented on reference/recognition.md:
```diff
@@ -2,11 +2,11 @@
+  SFInt32  maxObjects     -1      # {-1, [0, inf)}
+  SFBool   occlusion      TRUE    # {TRUE, FALSE}
+  SFColor  frameColor     1 0 0   # [0, 1]
+  SFInt32  frameThickness 1       # [0, inf]
```
wrong

omichel commented on reference/solid.md:
```diff
@@ -4,20 +4,20 @@ Derived from [Transform](transform.md).
+  SFFloat  translationStep     0.01          # [0, inf)
+  SFFloat  rotationStep        0.261799387   # [0, inf)
+  SFVec3f  linearVelocity      0 0 0         # [0, inf)
+  SFVec3f  angularVelocity     0 0 0         # [0, inf)
```
That's odd, why can't it be any vector?

omichel commented on reference/spotlight.md:
```diff
@@ -4,17 +4,17 @@ Derived from [Light](light.md).
+  SFFloat cutOffAngle      0.785398   # [0, pi/2)
+  SFVec3f direction        0 0 -1     # (-inf, inf)
+  SFFloat intensity        1          # [0, 1]
+  SFVec3f location         0 0 10     # (-inf, inf)
```
We have a problem with `SFVec3f`. Either we write "any vector" or something like "(-inf, inf) (-inf, inf) (-inf, inf)" or simply nothing. Similar problem with `SFVec2f` and `SFRotation`...

omichel commented on reference/touchsensor.md:
```diff
@@ -4,9 +4,9 @@ Derived from [Device](device.md).
-  SFFloat  resolution  -1
+  SFString type        "bumper"                  # ["bumper", "force", "force-3d"]
+  MFVec3f  lookupTable [ 0 0 0, 5000 50000 0 ]   # [see below]
+  SFFloat  resolution  -1                        # [-1, (0, inf)]
```
These 3 lines are wrong / non consistent.

tn12787 commented:
closer and closer still. I've fixed your comments and tried to improve `SFVec3f`, `SFVec2f` and `SFRotation` fields.

tn12787 commented:
👍

omichel commented on reference/physics.md:
```diff
@@ -2,11 +2,11 @@
+  SFFloat density       1000   # {-1, [0, inf)}
+  SFFloat mass          -1     # {-1, [0, inf)}
+  MFVec3f centerOfMass  [ ]    # {-1, [0, inf)}
+  MFVec3f inertiaMatrix [ ]    # {-1, [0, 1]}
```
Those two lines are not understandable to me...

omichel commented on reference/touchsensor.md:
```diff
@@ -4,9 +4,9 @@ Derived from [Device](device.md).
-  MFVec3f  lookupTable [ 0 0 0, 5000 50000 0 ]
-  SFFloat  resolution  -1
+  SFString type        "bumper"                  # {"bumper", "force", "force-3d"}
+  MFVec3f  lookupTable [ 0 0 0, 5000 50000 0 ]   # see below
```
Not consistent: you used "lookup table" previously.

omichel commented on reference/fluid.md:
```diff
@@ -4,15 +4,15 @@ Derived from [Transform](transform.md).
+  SFString description    ""        # any string
+  SFFloat  density        1000      # {-1, [0, inf)} 
+  SFFloat  viscosity      0.001     # [0, inf)
+  SFVec3f  streamVelocity 0 0 0     # any positive vector
```
Does it really have to be positive?

omichel commented on reference/jointparameters.md:
```diff
@@ -2,13 +2,13 @@
-  SFFloat dampingConstant 0     # damping constant (Ns/m or Nms)
-  SFFloat staticFriction  0     # friction constant (Ns/m or Nms)
+  SFFloat position        0       # [0, inf)
+  SFVec3f axis            0 0 1   # [0, 1]
```
&rarr; unit axis

omichel commented on reference/muscle.md:
```diff
@@ -2,12 +2,12 @@
-  SFBool   visible     TRUE
+  SFDouble maxRadius   0.2     # [0, inf)
+  SFVec3f  startOffset 0 0 0   # any positive vector
+  SFVec3f  endOffset   0 0 0   # any positive vector
```
Not sure they should be positive...

omichel commented on reference/propeller.md:
```diff
@@ -2,13 +2,13 @@
-  SFNode device           NULL  # RotationalMotor
-  SFNode fastHelix        NULL  # Solid node containing a graphical representation for rotation speed > 24 π rad/s (720 rpm)
-  SFNode slowHelix        NULL  # Solid node containing a graphical representation for rotation speed <= 24 π rad/s
+  SFVec3f shaftAxis       1 0 0   # [0, 1]
```
unit axis

omichel commented on reference/track.md:
```diff
@@ -4,10 +4,10 @@ Derived from [Solid](solid.md).
-  SFNode  animatedGeometry NULL
-  SFInt32 geometriesCount  10
+  MFNode  device           [ ]    # {LinearMotor, PositionSensor, Brake, PROTO}
+  SFVec2f textureAnimation 0 0    # any positive vector
```
Not sure it should be positive...

omichel commented on reference/worldinfo.md:
```diff
@@ -2,25 +2,25 @@
+  SFFloat  physicsDisableAngularThreshold 0.01      # [0, inf)
+  SFNode   defaultDamping                 NULL      # {Damping, PROTO}
+  SFFloat  inkEvaporation                 0         # [0, inf)
+  SFVec3f  northDirection                 1 0 0     # [-1, 1]
```
unit axis

tn12787 commented:
👍

omichel commented on reference/motor.md:
```diff
@@ -4,14 +4,14 @@ Derived from [Device](device.md).
+  SFFloat  maxPosition       0        # (-inf, inf) or [-pi, pi]
+  SFFloat  maxVelocity       10       # [0, inf)
+  SFString sound             ""       # any string
+  MFNode   muscles           []       # {Muscle (or derived PROTO)}
```
Why not `{Muscle, PROTO}` as done everywhere else?

omichel commented on reference/muscle.md:
```diff
@@ -2,12 +2,12 @@
+  SFDouble maxRadius   0.2     # [0, inf)
+  SFVec3f  startOffset 0 0 0   # any vector
+  SFVec3f  endOffset   0 0 0   # any vector
+  MFColor  colors      [ ]     # {SFColor, PROTO}
```
It's a list of RGB colors here, no PROTO allowed, SFColor is not a node!

omichel commented on reference/pen.md:
```diff
@@ -4,11 +4,11 @@ Derived from [Device](device.md).
-  SFFloat leadSize    0.002
-  SFFloat maxDistance 0.0   # >= 0.0
-  SFBool  write       TRUE
+  SFColor inkColor    0 0 0   # [0, 1]
```
Similar problem as with `SFVec3f`, we should should use "any color" here.

omichel commented on reference/propeller.md:
```diff
@@ -2,13 +2,13 @@
+  SFVec2f torqueConstants 1 0     # any vector
+  SFNode device           NULL    # {Solid (or derived)}
+  SFNode fastHelix        NULL    # {Solid (or derived)}
+  SFNode slowHelix        NULL    # {Solid (or derived)}
```
We should add `, PROTO` to be consistent with `endPoint`.

omichel commented on reference/lidar.md:
```diff
@@ -4,23 +4,23 @@ Derived from [Device](device.md).
+  SFFloat  defaultFrequency     10        # [minFrequency, maxFrequency]
+  SFFloat  minFrequency         1         # [0, maxFrequency)
+  SFFloat  maxFrequency         25        # [minFrequency, inf)
+  SFNode   rotatingHead         NULL      # {Solid (or derived)}
```
, PROTO

tn12787 commented:
I've fixed the last couple of comments and updated the color fields according to your suggestion. 

Sorry, I'm finding this PR hard to verify myself.

tn12787 commented:
👍

stefaniapedrazzi commented on reference/appearance.md:
```diff
@@ -2,11 +2,11 @@
-  MFNode   shaders          NULL
-  SFString name             "appearance"
+  SFNode   material         NULL           # {Material, PROTO}
+  SFNode   texture          NULL           # {ImageTexture, PROTO}
```
Missing "MultiTexture" and "ComposedCubeMapTexture"

stefaniapedrazzi commented on reference/hinge2joint.md:
```diff
@@ -4,9 +4,9 @@ Derived from [HingeJoint](hingejoint.md).
-  SFNode  jointParameters2 NULL # JointParameters for second axis
-  MFNode  device2          [ ]  # RotationalMotor, PositionSensor and Brake
-  SFFloat position2        0    # initial position with respect to the second hinge (rad)
+  SFNode  jointParameters2 NULL   # {JointParameters2, PROTO}
```
JointParameters2 doesn't exist -> JointParameters

stefaniapedrazzi commented on reference/joint.md:
```diff
@@ -2,8 +2,8 @@
 Joint {
-  SFNode jointParameters NULL # a joint parameters node
-  SFNode endPoint        NULL # Solid or SolidReference
+  SFNode jointParameters NULL   # {JointParameters, PROTO}
```
To be consistent JointParameters (or derived) -> depending on the type of joint JointParameters or HingeJointParameters is allowed

stefaniapedrazzi commented on reference/joint.md:
```diff
@@ -2,8 +2,8 @@
-  SFNode jointParameters NULL # a joint parameters node
-  SFNode endPoint        NULL # Solid or SolidReference
+  SFNode jointParameters NULL   # {JointParameters, PROTO}
+  SFNode endPoint        NULL   # {Solid, Slot, PROTO}
```
To be consistent Solid (or derived) and missing SolidReference node.

stefaniapedrazzi commented on reference/propeller.md:
```diff
@@ -2,13 +2,13 @@
+  SFVec3f centerOfThrust  0 0 0   # any vector
+  SFVec2f thrustConstants 1 0     # any vector
+  SFVec2f torqueConstants 1 0     # any vector
+  SFNode device           NULL    # {Solid (or derived), PROTO}
```
Wrong: only RotationalMotor is allowed -> {RotationalMotor, PROTO}

stefaniapedrazzi commented on reference/sliderjoint.md:
```diff
@@ -4,8 +4,8 @@ Derived from [Joint](joint.md).
 SliderJoint {
-  MFNode  device   [ ] # linear motor or linear position sensor
-  SFFloat position 0   # initial position (m)
+  MFNode  device   [ ]   # {Linear Motor, Linear PositionSensor, PROTO}
```
* To be consistent with the other descriptions (see Track/HingeJoint node for example):
  * "Linear Motor" -> "LinearMotor"
  * "Linear PositionSensor" -> "PositionSensor"


* Missing Brake

stefaniapedrazzi commented on reference/track.md:
```diff
@@ -4,10 +4,10 @@ Derived from [Solid](solid.md).
-  SFInt32 geometriesCount  10
+  MFNode  device           [ ]    # {LinearMotor, PositionSensor, Brake, PROTO}
+  SFVec2f textureAnimation 0 0    # any vector
+  SFNode  animatedGeometry NULL   # {Shape, Group, PROTO}
```
Missing Slot and Transform

stefaniapedrazzi commented on reference/slot.md:
```diff
@@ -2,8 +2,8 @@
-  SFString type     ""
-  SFNode   endPoint NULL
+  SFString type     ""     # any string
+  SFNode   endPoint NULL   # {Slot, Group (or derived), PROTO}
```
Missing PointLight, SpotLight, Shape

tn12787 replied:
I'm tempted to put # {node, PROTO} here to be consistent with group.md.

tn12787 commented:
👍 (at this point I feel like the boy who cried "wolf")
